File: libsignal/service/src/main/java/org/whispersystems/signalservice/internal/push/http/ResumableUploadSpec.java
package org.whispersystems.signalservice.internal.push.http;

import org.signal.protos.resumableuploads.ResumableUpload;
import org.whispersystems.signalservice.api.push.exceptions.ResumeLocationInvalidException;
import org.whispersystems.util.Base64;

import java.io.IOException;

import okio.ByteString;

public final class ResumableUploadSpec {

  private final byte[] secretKey;
  private final byte[] iv;

  private final String  cdnKey;
  private final Integer cdnNumber;
  private final String  resumeLocation;
  private final Long    expirationTimestamp;

  public ResumableUploadSpec(byte[] secretKey,
                             byte[] iv,
                             String cdnKey,
                             int cdnNumber,
                             String resumeLocation,
                             long expirationTimestamp)
  {
    this.secretKey           = secretKey;
    this.iv                  = iv;
    this.cdnKey              = cdnKey;
    this.cdnNumber           = cdnNumber;
    this.resumeLocation      = resumeLocation;
    this.expirationTimestamp = expirationTimestamp;
  }

  public byte[] getSecretKey() {
    return secretKey;
  }

  public byte[] getIV() {
    return iv;
  }

  public String getCdnKey() {
    return cdnKey;
  }

  public Integer getCdnNumber() {
    return cdnNumber;
  }

  public String getResumeLocation() {
    return resumeLocation;
  }

  public Long getExpirationTimestamp() {
    return expirationTimestamp;
  }

  public String serialize() {
    ResumableUpload.Builder builder = new ResumableUpload.Builder()
                                                         .secretKey(ByteString.of(getSecretKey()))
                                                         .iv(ByteString.of(getIV()))
                                                         .timeout(getExpirationTimestamp())
                                                         .cdnNumber(getCdnNumber())
                                                         .cdnKey(getCdnKey())
                                                         .location(getResumeLocation())
                                                         .timeout(getExpirationTimestamp());

    return Base64.encodeBytes(builder.build().encode());
  }

  public static ResumableUploadSpec deserialize(String serializedSpec) throws ResumeLocationInvalidException {
    if (serializedSpec == null) return null;

    try {
      ResumableUpload resumableUpload = ResumableUpload.ADAPTER.decode(Base64.decode(serializedSpec));

      return new ResumableUploadSpec(
          resumableUpload.secretKey.toByteArray(),
          resumableUpload.iv.toByteArray(),
          resumableUpload.cdnKey,
          resumableUpload.cdnNumber,
          resumableUpload.location,
          resumableUpload.timeout
      );
    } catch (IOException e) {
      throw new ResumeLocationInvalidException();
    }
  }
}

File: lib/libsignal-service/src/main/java/org/whispersystems/signalservice/internal/push/http/ResumableUploadSpec.kt
package org.whispersystems.signalservice.internal.push.http

import okio.ByteString.Companion.toByteString
import org.signal.core.util.Base64
import org.signal.protos.resumableuploads.ResumableUpload
import org.whispersystems.signalservice.api.push.exceptions.ResumeLocationInvalidException
import java.io.IOException

/**
 * Contains data around how to begin or resume an upload.
 * For given attachment, this data be saved and reused within the [expirationTimestamp] window.
 */
class ResumableUploadSpec(
  val attachmentKey: ByteArray,
  val attachmentIv: ByteArray,
  val cdnKey: String,
  val cdnNumber: Int,
  val resumeLocation: String,
  val expirationTimestamp: Long,
  val headers: Map<String, String>
) {
  fun toProto(): ResumableUpload {
    return ResumableUpload(
      secretKey = attachmentKey.toByteString(),
      iv = attachmentIv.toByteString(),
      timeout = expirationTimestamp,
      cdnNumber = cdnNumber,
      cdnKey = cdnKey,
      location = resumeLocation,
      headers = headers.entries.map { ResumableUpload.Header(key = it.key, value_ = it.value) }
    )
  }

  fun serialize(): String {
    return Base64.encodeWithPadding(toProto().encode())
  }

  companion object {
    @Throws(ResumeLocationInvalidException::class)
    fun deserialize(serializedSpec: String?): ResumableUploadSpec? {
      try {
        val resumableUpload = ResumableUpload.ADAPTER.decode(Base64.decode(serializedSpec!!))
        return from(resumableUpload)
      } catch (e: IOException) {
        throw ResumeLocationInvalidException()
      }
    }

    @Throws(ResumeLocationInvalidException::class)
    fun from(resumableUpload: ResumableUpload?): ResumableUploadSpec? {
      if (resumableUpload == null) {
        return null
      }

      val headers: MutableMap<String, String> = HashMap()
      for (header in resumableUpload.headers) {
        headers[header.key] = header.value_
      }

      return ResumableUploadSpec(
        attachmentKey = resumableUpload.secretKey.toByteArray(),
        attachmentIv = resumableUpload.iv.toByteArray(),
        cdnKey = resumableUpload.cdnKey,
        cdnNumber = resumableUpload.cdnNumber,
        resumeLocation = resumableUpload.location,
        expirationTimestamp = resumableUpload.timeout,
        headers = headers
      )
    }
  }
}
